### ISSUE TITLE
tjInitCompress throws error on iOS

### ISSUE THREAD

**Issue by konsnos** [libjpeg-turbo/libjpeg-turbo#247]
Hey again,

so I've compiled and ran the library successfully on Windows and Android, but it seems to throw an error on an iPad Mini 2, and iPhone X.

![init_error](https://user-images.githubusercontent.com/405674/41725250-04b8c2b2-7578-11e8-80fc-57716f58be3e.png)

I'm following the instructions from [ARMv8 (64-bit)](https://github.com/libjpeg-turbo/libjpeg-turbo/blob/master/BUILDING.md#armv8-64-bit).

The code seems to throw an error when I call the tjInitCompress().
` malloc: *** error for object 0x1021e07a8: pointer being freed was not allocated `

**Comment by dcommander** [comment]
Which version of libjpeg-turbo are you building?

**Comment by konsnos** [comment]
That would be 2.0.0.

**Comment by Vangos** [comment]
Same issue for me, too. Cannot use the iOS build of LibJPEG Turbo.

@dcommander Is there a recommended donation amount to fix this issue?

**Comment by dcommander** [comment]
I will fix it prior to the 2.0.0 release.  It will just take time to do so.  Be patient.  Testing and debugging libjpeg-turbo on iOS is not an easy thing to do.

**Comment by Vangos** [comment]
No worries. I have been using this framework a lot, so I would be glad to contribute. Thank you very much for your quick reply.

**Comment by dcommander** [comment]
All donations are always appreciated.

**Comment by dcommander** [comment]
Can you verify whether this problem also existed with previous versions of libjpeg-turbo?  I can't reproduce it, nor can I find any recent changes to the `tjInitCompress()` function that would explain it.  As far as I can tell, free() is never called in the body of that function.

Also, it seems that this error can occur due to issues that are outside of libjpeg-turbo's control:
https://stackoverflow.com/questions/12049058/malloc-error-for-object-0x165060-pointer-being-freed-was-not-allocated

**Comment by Vangos** [comment]
I am testing on Mac OS X High Sierra, targeting iOS 7-11. Here are my findings for version 2.

The following piece of code returns null:

```
DLLEXPORT tjhandle tjInitCompress(void)
{
   tjinstance *this = NULL;

   if ((this = (tjinstance *)malloc(sizeof(tjinstance))) == NULL)
   {
      snprintf(errStr, JMSG_LENGTH_MAX, "tjInitCompress(): Memory allocation failure");
      return NULL;
   }
   
   MEMZERO(this, sizeof(tjinstance));
   snprintf(this->errStr, JMSG_LENGTH_MAX, "No error");
   return _tjInitCompress(this);
}
```

_On Windows and Android, the above method is properly allocating the required space._

I have not yet tested version 1.5, but I’ll give it a try.

Have you been able to run version 2 on an iOS device? It’s always giving the same error on my iPhone and iPad devices.

**Comment by dcommander** [comment]
Yes, I tested it on an iOS device (jailbroken 5S.)

**Comment by dcommander** [comment]
If you can figure out exactly where the NULL pointer is coming from, that would be helpful. If tjInitCompress() is returning NULL, then check the error string.

**Comment by Vangos** [comment]
Thank you. I will check the error messages further and get back to you.

I would also like to see if I compiled the library incorrectly. Is there a precompiled iOS version of the library? I am checking the assemblies on SourceForge, but none of them seems compatible with iOS.

**Comment by dcommander** [comment]
The iOS and Mac versions are provided in the same library. Download and install the DMG.

**Comment by Vangos** [comment]
Just tried the official binaries of the DMG (iPhone & iPad devices). All of them are giving me the following error message:

```
JPEG parameter struct mismatch: library thinks size is 504, caller expects 520
```
![image](https://user-images.githubusercontent.com/562680/41901274-24205bd2-7939-11e8-9c4e-4e0070963c02.png)

**Comment by dcommander** [comment]
That error message is thrown by `jpeg_create_compress()`.  It means that there is a version mismatch between the jpeglib.h header and the libjpeg.a (or libjpeg.dylib) library.  However, I have no idea why that error would be thrown by `tjInitCompress()`.  I've tried both a clean iOS ARM v8 build, using our official build scripts (https://github.com/libjpeg-turbo/buildscripts/tree/master/setupscripts), and the official project binaries.  Everything works fine for me,

**Comment by Vangos** [comment]
Thanks! I am going to check this and update the ticket accordingly.

**Comment by Vangos** [comment]
I am building the projects using Unity3D. Unity is automatically including LibJPEG version 8. I am including the latest version of LibJPEG Turbo in my XCode project.

I assume that would be the mismatch.

**Comment by dcommander** [comment]
Probably so.  You might be able to pass `-DWITH_JPEG8=1` to the libjpeg-turbo build to get around the issue.  If that doesn't work, then I have no idea.

**Comment by dcommander** [comment]
Closing as WFM, as it seems likely that this is not a bug.

**Comment by Adam-VisualVocal** [comment]
I ran into a similar issue with my use of libjpeg-turbo and epeg. Unity has its own version of libjpeg and some of the symbols collide. In my case I was able to work around it by absorbing libjpeg-turbo completely into the epeg library thereby hiding the libjpeg-turbo symbols from Unity. Granted you might not be able to do something similar depending upon your use case.

Here's how I did it if you're curious.
iOS: https://github.com/VISUAL-VOCAL/epeg/commit/cfbb94bbea128765d1f1c6efa997fe731f42570d
Android: https://github.com/VISUAL-VOCAL/epeg/commit/507789bf3fcb6af815fe8608eed1cf6eb44b1134
Windows: https://github.com/VISUAL-VOCAL/epeg/commit/ae64516039ef05315fab443c0135249132c05817
